In [1]:
import json
import os
from pathlib import Path
from google.colab import userdata

# api key for importing Kaggle and downloading the datasets
os.environ["KAGGLE_KEY"] = userdata.get('KAGGLE_KEY')
os.environ["KAGGLE_USERNAME"] = userdata.get('KAGGLE_USERNAME')
# api key for the json object below
api_key = userdata.get('API_KEY')

# uses pathlib Path
kaggle_path = Path('/root/.kaggle')
os.makedirs(kaggle_path, exist_ok=True)

# opens file and dumps python dict to json object
with open (kaggle_path/'kaggle.json', 'w') as handl:
    json.dump(api_key,handl)

os.chmod(kaggle_path/'kaggle.json', 600)

In [2]:
import kaggle
! kaggle datasets download joelmott/chicago-traffic-crashes-may-2024/

Dataset URL: https://www.kaggle.com/datasets/joelmott/chicago-traffic-crashes-may-2024/versions/
License(s): CC0-1.0
100% 372M/372M [00:04<00:00, 126MB/s]
100% 372M/372M [00:04<00:00, 94.8MB/s]


In [3]:
!unzip /content/chicago-traffic-crashes-may-2024.zip

Archive:  /content/chicago-traffic-crashes-may-2024.zip
  inflating: CPD_traffic_crashes-crashes-20240501.csv  
  inflating: CPD_traffic_crashes-people-20240501.csv  
  inflating: CPD_traffic_crashes-vehicles-20240501.csv  
  inflating: trimmed_chicago_crashes_data.csv  


In [4]:
import pandas as pd
crashes_df = pd.read_csv('/content/CPD_traffic_crashes-crashes-20240501.csv')
crashes_df

CRASH_RECORD_ID CRASH_DATE_EST_I  \
0       6c1659069e9c6285a650e70d6f9b574ed5f64c12888479...              NaN   
1       5f54a59fcb087b12ae5b1acff96a3caf4f2d37e79f8db4...              NaN   
2       61fcb8c1eb522a6469b460e2134df3d15f82e81fd93e9c...              NaN   
3       004cd14d0303a9163aad69a2d7f341b7da2a8572b2ab33...              NaN   
4       a1d5f0ea90897745365a4cbb06cc60329a120d89753fac...              NaN   
...                                                   ...              ...   
828506  790a72494289861e0d12786c57659df0fcc3ff90df83a6...                Y   
828507  0c3ba5ebe2928b85f42ad69485a11977fef61a3a5b6b82...              NaN   
828508  54d55bfcc6627f587abbe0d14c42e51b812f930566fb06...              NaN   
828509  fcf3c5ddd8ba79a1b43c039a2507b201c999d659aa85d3...              NaN   
828510  f20453c5723af4ecc9c739665b2565af1529b4e705f63d...              NaN   

                    CRASH_DATE  POSTED_SPEED_LIMIT TRAFFIC_CONTROL_DEVICE  \
0       08/18/2023 12:50:00 PM                  15                  OTHER   
1       07/29/2023 02:45:00 PM                  30         TRAFFIC SIGNAL   
2       08/18/2023 05:58:00 PM                  30            NO CONTROLS   
3       11/26/2019 08:38:00 AM                  25            NO CONTROLS   
4       08/18/2023 10:45:00 AM                  20            NO CONTROLS   
...                        ...                 ...                    ...   
828506  09/08/2020 03:10:00 PM                  30            NO CONTROLS   
828507  08/18/2023 10:00:00 PM                  30         TRAFFIC SIGNAL   
828508  12/28/2019 01:16:00 AM                  35                UNKNOWN   
828509  04/12/2020 07:10:00 AM                  30            NO CONTROLS   
828510  05/17/2023 10:20:00 AM                  30            NO CONTROLS   

            DEVICE_CONDITION WEATHER_CONDITION      LIGHTING_CONDITION  \
0       FUNCTIONING PROPERLY             CLEAR                DAYLIGHT   
1       FUNCTIONING PROPERLY             CLEAR                DAYLIGHT   
2                NO CONTROLS             CLEAR                DAYLIGHT   
3                NO CONTROLS             CLEAR                DAYLIGHT   
4                NO CONTROLS             CLEAR                DAYLIGHT   
...                      ...               ...                     ...   
828506           NO CONTROLS              RAIN                    DUSK   
828507  FUNCTIONING PROPERLY             CLEAR  DARKNESS, LIGHTED ROAD   
828508               UNKNOWN             CLEAR  DARKNESS, LIGHTED ROAD   
828509           NO CONTROLS             CLEAR                DAYLIGHT   
828510           NO CONTROLS             CLEAR                DAYLIGHT   

                FIRST_CRASH_TYPE                  TRAFFICWAY_TYPE  ...  \
0                       REAR END                            OTHER  ...   
1           PARKED MOTOR VEHICLE  DIVIDED - W/MEDIAN (NOT RAISED)  ...   
2                   PEDALCYCLIST                      NOT DIVIDED  ...   
3                     PEDESTRIAN                          ONE-WAY  ...   
4                   FIXED OBJECT                            OTHER  ...   
...                          ...                              ...  ...   
828506      PARKED MOTOR VEHICLE                      NOT DIVIDED  ...   
828507                   TURNING                         FOUR WAY  ...   
828508      PARKED MOTOR VEHICLE                          ONE-WAY  ...   
828509  SIDESWIPE SAME DIRECTION  DIVIDED - W/MEDIAN (NOT RAISED)  ...   
828510      PARKED MOTOR VEHICLE                      NOT DIVIDED  ...   

        INJURIES_NON_INCAPACITATING INJURIES_REPORTED_NOT_EVIDENT  \
0                               1.0                           0.0   
1                               0.0                           0.0   
2                               1.0                           0.0   
3                               0.0                           0.0   
4                               0.0                      

In [5]:
#crashes_df = pd.read_csv('/content/drive/MyDrive/CPD_traffic_crashes-crashes-20240501.csv')
#crashes_df.head(3)

In [6]:
crashes_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 828511 entries, 0 to 828510
Data columns (total 48 columns):
 #   Column                         Non-Null Count   Dtype  
---  ------                         --------------   -----  
 0   CRASH_RECORD_ID                828511 non-null  object 
 1   CRASH_DATE_EST_I               61842 non-null   object 
 2   CRASH_DATE                     828511 non-null  object 
 3   POSTED_SPEED_LIMIT             828511 non-null  int64  
 4   TRAFFIC_CONTROL_DEVICE         828511 non-null  object 
 5   DEVICE_CONDITION               828511 non-null  object 
 6   WEATHER_CONDITION              828511 non-null  object 
 7   LIGHTING_CONDITION             828511 non-null  object 
 8   FIRST_CRASH_TYPE               828511 non-null  object 
 9   TRAFFICWAY_TYPE                828511 non-null  object 
 10  LANE_CNT                       199011 non-null  float64
 11  ALIGNMENT                      828511 non-null  object 
 12  ROADWAY_SURFACE_COND          

Keeping the project's prompt in mind, it looks like 'PRIM_CONTRIBUTORY_CAUSE' is our main target column.

As far as feature engineering goes, there are several different already-ohe'd categorical injury columns, maybe you add up all the ones with fatalities and bad injuries or something.

Next would be getting a feel for the column values where the data description isn't already obvious.

In [7]:
crashes_df['PRIM_CONTRIBUTORY_CAUSE'].value_counts()

PRIM_CONTRIBUTORY_CAUSE
UNABLE TO DETERMINE                                                                 322717
FAILING TO YIELD RIGHT-OF-WAY                                                        91120
FOLLOWING TOO CLOSELY                                                                80452
NOT APPLICABLE                                                                       43885
IMPROPER OVERTAKING/PASSING                                                          40874
FAILING TO REDUCE SPEED TO AVOID CRASH                                               34961
IMPROPER BACKING                                                                     32492
IMPROPER LANE USAGE                                                                  29635
DRIVING SKILLS/KNOWLEDGE/EXPERIENCE                                                  27817
IMPROPER TURNING/NO SIGNAL                                                           27539
DISREGARDING TRAFFIC SIGNALS                                      

In [8]:
len(crashes_df['PRIM_CONTRIBUTORY_CAUSE'].value_counts())

40

40 classes?!? Some of them occured fewer than 100 times! I guess I'd have to drop the 'UNABLE TO DETERMINE' and 'NOT APPLICABLE' rows. Then, I think I'd need to group the remaining 38 classes together somehow.

I'll work on that off-and-on while I continue to check out the remaining features in this crashes DataFrame. Let's start with 'TRAFFIC_CONTROL_DEVICE':

In [9]:
crashes_df['TRAFFIC_CONTROL_DEVICE'].value_counts()

TRAFFIC_CONTROL_DEVICE
NO CONTROLS                 470741
TRAFFIC SIGNAL              229645
STOP SIGN/FLASHER            82248
UNKNOWN                      33447
OTHER                         5623
LANE USE MARKING              1226
YIELD                         1218
OTHER REG. SIGN                932
OTHER WARNING SIGN             682
RAILROAD CROSSING GATE         536
PEDESTRIAN CROSSING SIGN       532
FLASHING CONTROL SIGNAL        334
DELINEATORS                    331
SCHOOL ZONE                    315
POLICE/FLAGMAN                 279
OTHER RAILROAD CROSSING        182
RR CROSSING SIGN               164
NO PASSING                      51
BICYCLE CROSSING SIGN           25
Name: count, dtype: int64

This feature seems to be almost enitrely made up of the classes 'no controls', 'traffic signal', 'stop sign/flasher', or 'unknown.' The rest are quite rare, relatively speaking. I think this could be a helpful column, but the distribution is really skewed, so in the interest of trimming features that may add more noise to our model when model interpretability is a priority, let's drop this one.

Let's move on to the 'DEVICE_CONDITION' feature:

In [10]:
crashes_df['DEVICE_CONDITION'].value_counts()

DEVICE_CONDITION
NO CONTROLS                 476188
FUNCTIONING PROPERLY        283564
UNKNOWN                      55753
OTHER                         6336
FUNCTIONING IMPROPERLY        3890
NOT FUNCTIONING               2397
WORN REFLECTIVE MATERIAL       287
MISSING                         96
Name: count, dtype: int64

I had to look it up, but 'NO CONTROLS' refers to a traffic control device that doesn't change like a light would, but rather is a permanent fixture like a sign or road markings. I guess it's not surprising that the vast majority are in this category and then another huge chunk are at intersections where the light was working properly. Note a decent amount of 'UNKNOWN' entries here. With only 3,890 functioning improperly compared to 759,752 either not having controls or working normally, I don't think this feature will be very helpful, so let's drop it.

Next to check out is 'FIRST_CRASH_TYPE':

In [11]:
crashes_df['FIRST_CRASH_TYPE'].value_counts()

FIRST_CRASH_TYPE
PARKED MOTOR VEHICLE            192428
REAR END                        184855
SIDESWIPE SAME DIRECTION        126534
TURNING                         118537
ANGLE                            90173
FIXED OBJECT                     38930
PEDESTRIAN                       19330
PEDALCYCLIST                     12448
SIDESWIPE OPPOSITE DIRECTION     11666
OTHER OBJECT                      8216
REAR TO FRONT                     8094
HEAD ON                           7061
REAR TO SIDE                      4807
OTHER NONCOLLISION                2606
REAR TO REAR                      1676
ANIMAL                             596
OVERTURNED                         508
TRAIN                               46
Name: count, dtype: int64

I can't help but wonder whether this is a good *predictor* for an accident's primary cause when it describes the instance of the accident, not what happens beforehand. It turns out that, later on, we'll find a better column for this in the vehicles DataFrame ('MANEUVER') that will describe the action that led to the accident, so let's drop this column in favor of that one.

The following 'TRAFFICWAY_TYPE' column looks useful as well, but it has so many classes that, along with the other useful features we'll eventually find, it'll wind up adding a lot of noise to our data with the sparsity that will result once it's one-hot encoded.

In [12]:
crashes_df['TRAFFICWAY_TYPE'].value_counts()

TRAFFICWAY_TYPE
NOT DIVIDED                        359452
DIVIDED - W/MEDIAN (NOT RAISED)    132085
ONE-WAY                            105558
PARKING LOT                         56368
FOUR WAY                            53091
DIVIDED - W/MEDIAN BARRIER          47287
OTHER                               22646
ALLEY                               13690
T-INTERSECTION                      10660
UNKNOWN                              9656
CENTER TURN LANE                     5978
DRIVEWAY                             2691
RAMP                                 2595
UNKNOWN INTERSECTION TYPE            2455
Y-INTERSECTION                       1164
FIVE POINT, OR MORE                  1161
TRAFFIC ROUTE                         948
NOT REPORTED                          592
ROUNDABOUT                            262
L-INTERSECTION                        172
Name: count, dtype: int64

'ALIGNMENT' has a distribution issue with the top class making up almost 98% of this feature, so it won't be too helpful.

In [13]:
crashes_df['ALIGNMENT'].value_counts()

ALIGNMENT
STRAIGHT AND LEVEL       808482
STRAIGHT ON GRADE         10279
CURVE, LEVEL               5971
STRAIGHT ON HILLCREST      2169
CURVE ON GRADE             1243
CURVE ON HILLCREST          367
Name: count, dtype: int64

'CRASH_TYPE' would make for a helpful target column if we were interested in less serious versus more serious accidents, but we're looking into the cause of *all* accidents and this column ultimately describes the aftermath, not the cause, so we'll drop for this particular use case.

In [14]:
crashes_df['CRASH_TYPE'].value_counts()

CRASH_TYPE
NO INJURY / DRIVE AWAY              607050
INJURY AND / OR TOW DUE TO CRASH    221461
Name: count, dtype: int64

The following column covers whether a right-of-way violation occured, but that's in our target variable, so we should eliminate this one.

In [15]:
crashes_df['NOT_RIGHT_OF_WAY_I'].value_counts()

NOT_RIGHT_OF_WAY_I
Y    34674
N     3493
Name: count, dtype: int64

The 'DAMAGE' column refers to the aftermath, not the cause, so this can be dropped.

In [16]:
crashes_df['DAMAGE'].value_counts()

DAMAGE
OVER $1,500      515387
$501 - $1,500    218754
$500 OR LESS      94370
Name: count, dtype: int64

These 'BEAT_OF_OCCURANCE' classes are categorical even if they're listed here as numbers. Subsequently, they would need to be one-hot encoded, and there are just far too many of them for that, so we'll have to drop this column as well.

In [17]:
crashes_df['BEAT_OF_OCCURRENCE'].value_counts()

BEAT_OF_OCCURRENCE
1834.0    10067
114.0      8490
813.0      8332
815.0      7899
1831.0     7678
          ...  
1653.0      430
1655.0      260
1652.0      200
1650.0       69
6100.0        4
Name: count, Length: 276, dtype: int64

Now that I have a sense of what all these columns mean, it's time to consider which should be eliminated to help reduce dimensionality. This is especially important since so many of these columns will have to be one-hot encoded.

Since the goal of this project is to predict the *primary cause* of an accident, we might consider dropping those columns that inherently have nothing to do with a potential cause (such as 'DATE_POLICE_NOTIFIED', 'STREET NO', 'STREET NAME', 'PHOTOS TAKEN', 'STATEMENTS TAKEN', 'REPORT_TYPE') and/or only pertain to the effect (all the damage, injury, and fatality columns). We can also drop the 'SEC_CONTRIBUTORY_CAUSE' or we'll probably have a lot of multicollinearity issues.

There are a few redundant columns for our specific use case. 'LOCATION' just puts the latitude & longitude columns together. Since those two columns will provide location data *and* 'BEAT_OF_OCCURRENCE' has 276 categorical values, let's eliminate the beat column as well. Also, 'CRASH_DATE_EST_I' and 'CRASH_DATE' are broken down later via 'CRASH_HOUR', 'CRASH_DAY_OF_WEEK', and 'CRASH_MONTH'.

Finally, some columns are woefully empty; 'LANE_CNT', 'INTERSECTION_RELATED_I', 'NOT_RIGHT_OF_WAY_I', 'HIT_AND_RUN_I', 'DOORING_I', and all the work zone columns only have somewhere between 68.7-99% empty, which is grounds for dropping.

I'll keep 'CRASH_RECORD_ID' for now so I can join the 'people' & 'vehicles' tables later.

In [18]:
crashes_columns_to_drop = ['CRASH_DATE', 'CRASH_DATE_EST_I', 'DEVICE_CONDITION', 'FIRST_CRASH_TYPE', 'CRASH_TYPE',
                           'REPORT_TYPE', 'DAMAGE', 'DATE_POLICE_NOTIFIED', 'SEC_CONTRIBUTORY_CAUSE',
                           'TRAFFIC_CONTROL_DEVICE', 'STREET_NO', 'STREET_NAME', 'BEAT_OF_OCCURRENCE', 'PHOTOS_TAKEN_I',
                           'STATEMENTS_TAKEN_I', 'MOST_SEVERE_INJURY', 'INJURIES_TOTAL', 'INJURIES_FATAL',
                           'INJURIES_INCAPACITATING', 'INJURIES_NON_INCAPACITATING', 'INJURIES_REPORTED_NOT_EVIDENT',
                           'INJURIES_NO_INDICATION', 'INJURIES_UNKNOWN', 'LOCATION', 'LANE_CNT', 'ALIGNMENT',
                           'INTERSECTION_RELATED_I', 'NOT_RIGHT_OF_WAY_I', 'HIT_AND_RUN_I', 'DOORING_I',
                           'WORK_ZONE_I', 'WORK_ZONE_TYPE', 'WORKERS_PRESENT_I', 'TRAFFICWAY_TYPE']

In [19]:
crashes_df_refined = crashes_df.drop(crashes_columns_to_drop, axis=1)
crashes_df_refined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 828511 entries, 0 to 828510
Data columns (total 14 columns):
 #   Column                   Non-Null Count   Dtype  
---  ------                   --------------   -----  
 0   CRASH_RECORD_ID          828511 non-null  object 
 1   POSTED_SPEED_LIMIT       828511 non-null  int64  
 2   WEATHER_CONDITION        828511 non-null  object 
 3   LIGHTING_CONDITION       828511 non-null  object 
 4   ROADWAY_SURFACE_COND     828511 non-null  object 
 5   ROAD_DEFECT              828511 non-null  object 
 6   PRIM_CONTRIBUTORY_CAUSE  828511 non-null  object 
 7   STREET_DIRECTION         828507 non-null  object 
 8   NUM_UNITS                828511 non-null  int64  
 9   CRASH_HOUR               828511 non-null  int64  
 10  CRASH_DAY_OF_WEEK        828511 non-null  int64  
 11  CRASH_MONTH              828511 non-null  int64  
 12  LATITUDE                 822769 non-null  float64
 13  LONGITUDE                822769 non-null  float64
dtypes: f

Now let's take care of the all the null values. The NaN's in 'LATITUDE' and 'LONGITUDE' amount to 0.007% and there are only four NaN's in 'STREET_DIRECTION' and wait, I should only do this once I join the other tables or we'll have missing record problems, so for now, let's one-hot encode the categorical columns.

Since this is a large dataset, I'll use a train-test-split as my validation strategy. In order to prevent overfitting, I'll use some form of regularization (I'll try different kinds and use what works best). Subsequently, I will not need to drop any columns when using one-hot encoding because regularization will take care of any singularity issues among the categorical variables BUT nevermind, **I don't ohe until after the tts.**

In [20]:
people_df = pd.read_csv('/content/CPD_traffic_crashes-people-20240501.csv')
people_df.head(3)

<ipython-input-20-26241325ca0e>:1: DtypeWarning: Columns (19,23,24,25,28) have mixed types. Specify dtype option on import or set low_memory=False.
  people_df = pd.read_csv('/content/CPD_traffic_crashes-people-20240501.csv')


PERSON_ID PERSON_TYPE                                    CRASH_RECORD_ID  \
0   O749947      DRIVER  81dc0de2ed92aa62baccab641fa377be7feb1cc47e6554...   
1   O871921      DRIVER  af84fb5c8d996fcd3aefd36593c3a02e6e7509eeb27568...   
2    O10018      DRIVER  71162af7bf22799b776547132ebf134b5b438dcf3dac6b...   

   VEHICLE_ID              CRASH_DATE  SEAT_NO     CITY STATE ZIPCODE SEX  \
0    834816.0  09/28/2019 03:30:00 AM      NaN  CHICAGO    IL   60651   M   
1    827212.0  04/13/2020 10:50:00 PM      NaN  CHICAGO    IL   60620   M   
2      9579.0  11/01/2015 05:00:00 AM      NaN      NaN   NaN     NaN   X   

   ...  EMS_RUN_NO     DRIVER_ACTION DRIVER_VISION PHYSICAL_CONDITION  \
0  ...         NaN           UNKNOWN       UNKNOWN            UNKNOWN   
1  ...         NaN              NONE  NOT OBSCURED             NORMAL   
2  ...         NaN  IMPROPER BACKING       UNKNOWN            UNKNOWN   

  PEDPEDAL_ACTION PEDPEDAL_VISIBILITY PEDPEDAL_LOCATION        BAC_RESULT  \
0             NaN                 NaN               NaN  TEST NOT OFFERED   
1             NaN                 NaN               NaN  TEST NOT OFFERED   
2             NaN                 NaN               NaN  TEST NOT OFFERED   

  BAC_RESULT VALUE CELL_PHONE_USE  
0              NaN            NaN  
1              NaN            NaN  
2              NaN            NaN  

[3 rows x 29 columns]

In [21]:
people_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1818705 entries, 0 to 1818704
Data columns (total 29 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   PERSON_ID              object 
 1   PERSON_TYPE            object 
 2   CRASH_RECORD_ID        object 
 3   VEHICLE_ID             float64
 4   CRASH_DATE             object 
 5   SEAT_NO                float64
 6   CITY                   object 
 7   STATE                  object 
 8   ZIPCODE                object 
 9   SEX                    object 
 10  AGE                    float64
 11  DRIVERS_LICENSE_STATE  object 
 12  DRIVERS_LICENSE_CLASS  object 
 13  SAFETY_EQUIPMENT       object 
 14  AIRBAG_DEPLOYED        object 
 15  EJECTION               object 
 16  INJURY_CLASSIFICATION  object 
 17  HOSPITAL               object 
 18  EMS_AGENCY             object 
 19  EMS_RUN_NO             object 
 20  DRIVER_ACTION          object 
 21  DRIVER_VISION          object 
 22  PHYSICAL_CONDITION

As with the crashes set, our first task is to understand and then reduce the columns we'll use. Specifically, we'll determine & eliminate those that don't necessarily pertain to the the *cause* of an accident, but rather the aftermath. In order to decrease dimensionality, we'll also drop categorical columns that have high cardinality and aren't likely to contribute much, like zipcodes or the state from which the drivers license was issued.

In [22]:
people_df['PERSON_TYPE'].value_counts()

PERSON_TYPE
DRIVER                 1415231
PASSENGER               367648
PEDESTRIAN               21244
BICYCLE                  12794
NON-MOTOR VEHICLE         1482
NON-CONTACT VEHICLE        306
Name: count, dtype: int64

In [23]:
with pd.option_context('display.max_rows', None, 'display.max_columns', None):
    print (people_df['DRIVER_ACTION'].value_counts())

DRIVER_ACTION
NONE                                 517295
UNKNOWN                              365078
FAILED TO YIELD                      131914
OTHER                                129661
FOLLOWED TOO CLOSELY                  85675
IMPROPER BACKING                      42948
IMPROPER TURN                         37980
IMPROPER LANE CHANGE                  37173
IMPROPER PASSING                      32369
DISREGARDED CONTROL DEVICES           25564
TOO FAST FOR CONDITIONS               21734
WRONG WAY/SIDE                         5813
IMPROPER PARKING                       5329
OVERCORRECTED                          2710
EVADING POLICE VEHICLE                 2304
CELL PHONE USE OTHER THAN TEXTING      2126
EMERGENCY VEHICLE ON CALL              1328
TEXTING                                 569
STOPPED SCHOOL BUS                      174
LICENSE RESTRICTIONS                     61
Name: count, dtype: int64


In [24]:
people_df['DRIVER_VISION'].value_counts()

DRIVER_VISION
NOT OBSCURED              731001
UNKNOWN                   681356
OTHER                      14386
MOVING VEHICLES             8198
PARKED VEHICLES             5044
WINDSHIELD (WATER/ICE)      3987
BLINDED - SUNLIGHT          1716
TREES, PLANTS                578
BUILDINGS                    516
BLINDED - HEADLIGHTS         155
BLOWING MATERIALS            102
HILLCREST                     96
EMBANKMENT                    83
SIGNBOARD                     35
Name: count, dtype: int64

Safety equipment almost certainly has an impact on injuries & fatalities, but if we're prioritizing an effort to minimize dimensions, then weighing whether they have an impact on the *cause* probably means we can drop this column.

'DRIVER_ACTION' can be dropped since it's pretty redundant when it comes to our target. 'DRIVER_VISION' categorizes whether and how the driver's vision was impeded, but its classes are almost enitrely either 'unobscured' or 'unknown', so I don't think it will helpful here. The same goes for 'PHYSICAL_CONDITION' where only a handful aren't either 'normal' or 'unknown'.

'PEDPEDAL_ACTION' refers to what a pedestrian or cyclist was doing if they were involved in the accident. This would be useful if we were just looking at such accidents, but we are investigating all traffic crashes. Let's see the extent to which these ped/pedal accidents make up the total:

In [25]:
pedpedal_action_vc = people_df['PEDPEDAL_ACTION'].value_counts()
pedpedal_action_vc_sum = sum(pedpedal_action_vc)
pedpedal_action_vc_sum / people_df.shape[0]

0.01903057395234521

It looks like this would only apply to less than 2% of the total rows, so it will only really add noise, so let's drop it along with 'PEDPEDAL_VISIBILITY' and 'PEDPEDAL_LOCATION'.

In [26]:
people_df['BAC_RESULT'].value_counts()

BAC_RESULT
TEST NOT OFFERED                   1427625
TEST REFUSED                         15087
TEST PERFORMED, RESULTS UNKNOWN       3496
TEST TAKEN                            2576
Name: count, dtype: int64

'BAC_RESULT' would be helpful but only a tiny percentage involve any tests taken, so it will not help here.

In [27]:
people_df['CELL_PHONE_USE'].value_counts()

CELL_PHONE_USE
Y    752
N    407
Name: count, dtype: int64

Unfortunately, 'CELL_PHONE_USE' doesn't show enough entries to be useful, either. With that, we have at least a good starting list of columns to drop, so let's drop them and move on to the last part of this dataset.

In [28]:
people_columns_to_drop = ['PERSON_ID', 'VEHICLE_ID', 'CRASH_DATE', 'SEAT_NO', 'STATE', 'ZIPCODE', 'DRIVERS_LICENSE_STATE',
                   'SAFETY_EQUIPMENT', 'AIRBAG_DEPLOYED', 'EJECTION', 'INJURY_CLASSIFICATION', 'HOSPITAL',
                   'EMS_AGENCY', 'EMS_RUN_NO', 'DRIVER_ACTION', 'DRIVER_VISION', 'PHYSICAL_CONDITION',
                   'PEDPEDAL_ACTION', 'PEDPEDAL_VISIBILITY', 'PEDPEDAL_LOCATION', 'BAC_RESULT',
                   'BAC_RESULT VALUE', 'CELL_PHONE_USE']

In [29]:
people_df_refined = people_df.drop(people_columns_to_drop, axis=1)
people_df_refined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1818705 entries, 0 to 1818704
Data columns (total 6 columns):
 #   Column                 Dtype  
---  ------                 -----  
 0   PERSON_TYPE            object 
 1   CRASH_RECORD_ID        object 
 2   CITY                   object 
 3   SEX                    object 
 4   AGE                    float64
 5   DRIVERS_LICENSE_CLASS  object 
dtypes: float64(1), object(5)
memory usage: 83.3+ MB


Finally, let's examine the third part of this dataset on vehicles:

In [30]:
vehicles_df = pd.read_csv('/content/CPD_traffic_crashes-vehicles-20240501.csv')
vehicles_df.head(3)

<ipython-input-30-04261ed8d1c9>:1: DtypeWarning: Columns (20,39,40,41,43,47,48,49,52,54,57,58,60,70) have mixed types. Specify dtype option on import or set low_memory=False.
  vehicles_df = pd.read_csv('/content/CPD_traffic_crashes-vehicles-20240501.csv')


CRASH_UNIT_ID                                    CRASH_RECORD_ID  \
0        1717556  7b1763088507f77e0e552c009a6bf89a4d6330c7527706...   
1        1717574  2603ff5a88f0b9b54576934c5ed4e4a64e8278e005687b...   
2        1717579  a52ef70e33d468b855b5be44e8638a564434dcf99c0edf...   

               CRASH_DATE  UNIT_NO UNIT_TYPE  NUM_PASSENGERS  VEHICLE_ID  \
0  12/06/2023 03:24:00 PM        1    DRIVER             NaN   1634931.0   
1  12/06/2023 04:00:00 PM        2    DRIVER             NaN   1634978.0   
2  12/06/2023 04:30:00 PM        1    DRIVER             NaN   1634948.0   

  CMRC_VEH_I      MAKE    MODEL  ... TRAILER1_LENGTH  TRAILER2_LENGTH  \
0        NaN    NISSAN   SENTRA  ...             NaN              NaN   
1        NaN  CHRYSLER  SEBRING  ...             NaN              NaN   
2        NaN    SUBARU  OUTBACK  ...             NaN              NaN   

  TOTAL_VEHICLE_LENGTH AXLE_CNT VEHICLE_CONFIG CARGO_BODY_TYPE LOAD_TYPE  \
0                  NaN      NaN            NaN             NaN       NaN   
1                  NaN      NaN            NaN             NaN       NaN   
2                  NaN      NaN            NaN             NaN       NaN   

  HAZMAT_OUT_OF_SERVICE_I MCS_OUT_OF_SERVICE_I  HAZMAT_CLASS  
0                     NaN                  NaN           NaN  
1                     NaN                  NaN           NaN  
2                     NaN                  NaN           NaN  

[3 rows x 71 columns]

In [31]:
vehicles_df.shape

(1690173, 71)

In [32]:
vehicles_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1690173 entries, 0 to 1690172
Data columns (total 71 columns):
 #   Column                    Non-Null Count    Dtype  
---  ------                    --------------    -----  
 0   CRASH_UNIT_ID             1690173 non-null  int64  
 1   CRASH_RECORD_ID           1690173 non-null  object 
 2   CRASH_DATE                1690173 non-null  object 
 3   UNIT_NO                   1690173 non-null  int64  
 4   UNIT_TYPE                 1688017 non-null  object 
 5   NUM_PASSENGERS            250022 non-null   float64
 6   VEHICLE_ID                1651514 non-null  float64
 7   CMRC_VEH_I                31452 non-null    object 
 8   MAKE                      1651509 non-null  object 
 9   MODEL                     1651364 non-null  object 
 10  LIC_PLATE_STATE           1500227 non-null  object 
 11  VEHICLE_YEAR              1386705 non-null  float64
 12  VEHICLE_DEFECT            1651514 non-null  object 
 13  VEHICLE_TYPE              1

The first column that doesn't look too redundant is 'UNIT_TYPE':

In [33]:
vehicles_df['UNIT_TYPE'].value_counts()

UNIT_TYPE
DRIVER                 1415231
PARKED                  222855
PEDESTRIAN               21244
DRIVERLESS               13847
BICYCLE                  12794
NON-MOTOR VEHICLE         1482
NON-CONTACT VEHICLE        306
DISABLED VEHICLE           251
EQUESTRIAN                   7
Name: count, dtype: int64

Almost everything here is either a driver-driven vehicle or a parked one, so I'll probabl drop this one because I think we may find a more useful designation once we get to 'VEHICLE_TYPE'. Anyway, I hope those seven horses in the 'EQUESTRIAN' class were okay.

The next one to consider would be 'MAKE', but it turns out to have *way* too many different values and the resulting sparsity from the one-hot encoded columns would just add a lot of noise to our model.

In [34]:
len(vehicles_df['MAKE'].value_counts())

1335

I'm sure the 'MODEL' column would be even worse in this regard, so let's drop them both and search for primary causes elsewhere in the name of model interpretability.

In [35]:
vehicles_df['VEHICLE_DEFECT'].value_counts()

VEHICLE_DEFECT
NONE                892456
UNKNOWN             741379
OTHER                 9406
BRAKES                5177
TIRES                  839
STEERING               748
WHEELS                 419
SUSPENSION             269
FUEL SYSTEM            233
ENGINE/MOTOR           210
WINDOWS                111
LIGHTS                 101
CARGO                   57
SIGNALS                 42
RESTRAINT SYSTEM        25
TRAILER COUPLING        23
EXHAUST                 19
Name: count, dtype: int64

With almost every class being 'none' or 'unknown', this column won't help.

In [36]:
vehicles_df['VEHICLE_TYPE'].value_counts()

VEHICLE_TYPE
PASSENGER                                 1038710
SPORT UTILITY VEHICLE (SUV)                226692
UNKNOWN/NA                                 152665
VAN/MINI-VAN                                78278
PICKUP                                      53326
TRUCK - SINGLE UNIT                         31040
OTHER                                       19918
BUS OVER 15 PASS.                           17739
TRACTOR W/ SEMI-TRAILER                     15661
BUS UP TO 15 PASS.                           4659
MOTORCYCLE (OVER 150CC)                      3869
SINGLE UNIT TRUCK WITH TRAILER               2734
OTHER VEHICLE WITH TRAILER                   2253
TRACTOR W/O SEMI-TRAILER                     2069
AUTOCYCLE                                     669
MOPED OR MOTORIZED BICYCLE                    544
MOTOR DRIVEN CYCLE                            328
ALL-TERRAIN VEHICLE (ATV)                     182
FARM EQUIPMENT                                 84
3-WHEELED MOTORCYCLE (2 REAR WHEELS) 

With 60% of new-vhecile registrations in the US being SUVs in 2023 and this dataset only going back to 2015, I find it hard to believe that SUV's are outnumbered at almost 4-to-1 in this count. I'm getting worried about this dataset's viability.

For the sake of brevity, 'VEHICLE_USE' is all either 'personal' or 'unknown' or 'not in use', so it won't be of much help.

'TRAVEL_DIRECTION' may prove useful, but the cardinal directions (NSEW) are all *much* more frequent than the intermediate ones, so we'll just keep those rows with cardinals.

'MANEUVER' seems to be a really useful column when it comes to the primary cause of an accident, so we'll keep it, but we'll drop the rows that are 'unknown', 'other', or occured less than 1,000 times.

In [37]:
vehicles_df['MANEUVER'].value_counts()

MANEUVER
STRAIGHT AHEAD                        761499
PARKED                                227171
UNKNOWN/NA                            126724
SLOW/STOP IN TRAFFIC                  122016
TURNING LEFT                           98255
BACKING                                66781
TURNING RIGHT                          54715
PASSING/OVERTAKING                     40106
CHANGING LANES                         31673
OTHER                                  27722
ENTERING TRAFFIC LANE FROM PARKING     19410
MERGING                                11451
U-TURN                                  9411
STARTING IN TRAFFIC                     9341
LEAVING TRAFFIC LANE TO PARK            7925
AVOIDING VEHICLES/OBJECTS               6988
SKIDDING/CONTROL LOSS                   6296
ENTER FROM DRIVE/ALLEY                  6001
PARKED IN TRAFFIC LANE                  5091
SLOW/STOP - LEFT TURN                   3042
DRIVING WRONG WAY                       2452
NEGOTIATING A CURVE                     2077
S

Let's also examine 'FIRST_CONTACT_POINT':

In [38]:
vehicles_df['FIRST_CONTACT_POINT'].value_counts()

FIRST_CONTACT_POINT
FRONT                 323891
REAR                  218038
UNKNOWN               157277
SIDE-LEFT             113074
SIDE-RIGHT            107560
FRONT-LEFT-CORNER     104177
FRONT-RIGHT-CORNER    102005
FRONT-LEFT             81524
FRONT-RIGHT            76903
REAR-LEFT              68088
REAR-LEFT-CORNER       46318
OTHER                  41017
REAR-RIGHT             36488
REAR-RIGHT-CORNER      33782
TOTAL (ALL AREAS)      27475
SIDE-LEFT-REAR         26325
SIDE-RIGHT-REAR        19947
SIDE-LEFT-FRONT        17038
SIDE-RIGHT-FRONT       14665
NONE                   13141
ROOF                   11958
UNDER CARRIAGE          5728
TOP                     1970
Name: count, dtype: int64

This would be useful, but it's a little redundant and lengthy when compared to the crashes_df column 'FIRST_CRASH_TYPE.' The remaining columns are all too filled with NaN's, so we now have a short list of columns to keep:



In [39]:
vehicles_df_columns_to_keep = ['CRASH_RECORD_ID', 'TRAVEL_DIRECTION', 'MANEUVER']

In [40]:
vehicles_df_refined = vehicles_df[vehicles_df_columns_to_keep]
vehicles_df_refined.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1690173 entries, 0 to 1690172
Data columns (total 3 columns):
 #   Column            Non-Null Count    Dtype 
---  ------            --------------    ----- 
 0   CRASH_RECORD_ID   1690173 non-null  object
 1   TRAVEL_DIRECTION  1651514 non-null  object
 2   MANEUVER          1651514 non-null  object
dtypes: object(3)
memory usage: 38.7+ MB


Now we can merge these more managable, concise DataFrames together and then begin removing the rows mentioned above.

In [41]:
#First, we'll merge the crashes & people df's:
merged_crashes = pd.merge(left=crashes_df_refined,
                          right=people_df_refined,
                          left_on='CRASH_RECORD_ID',
                          right_on='CRASH_RECORD_ID')

#Now the newly merged df and the vehicles one:
merged_crashes = pd.merge(left=merged_crashes,
                          right=vehicles_df_refined,
                          left_on='CRASH_RECORD_ID',
                          right_on='CRASH_RECORD_ID')

merged_crashes.head()

CRASH_RECORD_ID  POSTED_SPEED_LIMIT  \
0  6c1659069e9c6285a650e70d6f9b574ed5f64c12888479...                  15   
1  6c1659069e9c6285a650e70d6f9b574ed5f64c12888479...                  15   
2  6c1659069e9c6285a650e70d6f9b574ed5f64c12888479...                  15   
3  6c1659069e9c6285a650e70d6f9b574ed5f64c12888479...                  15   
4  5f54a59fcb087b12ae5b1acff96a3caf4f2d37e79f8db4...                  30   

  WEATHER_CONDITION LIGHTING_CONDITION ROADWAY_SURFACE_COND ROAD_DEFECT  \
0             CLEAR           DAYLIGHT                  DRY  NO DEFECTS   
1             CLEAR           DAYLIGHT                  DRY  NO DEFECTS   
2             CLEAR           DAYLIGHT                  DRY  NO DEFECTS   
3             CLEAR           DAYLIGHT                  DRY  NO DEFECTS   
4             CLEAR           DAYLIGHT                  DRY  NO DEFECTS   

                  PRIM_CONTRIBUTORY_CAUSE STREET_DIRECTION  NUM_UNITS  \
0                   FOLLOWING TOO CLOSELY                W          2   
1                   FOLLOWING TOO CLOSELY                W          2   
2                   FOLLOWING TOO CLOSELY                W          2   
3                   FOLLOWING TOO CLOSELY                W          2   
4  FAILING TO REDUCE SPEED TO AVOID CRASH                S          4   

   CRASH_HOUR  ...  CRASH_MONTH  LATITUDE  LONGITUDE  PERSON_TYPE  \
0          12  ...            8       NaN        NaN       DRIVER   
1          12  ...            8       NaN        NaN       DRIVER   
2          12  ...            8       NaN        NaN       DRIVER   
3          12  ...            8       NaN        NaN       DRIVER   
4          14  ...            7  41.85412 -87.665902       DRIVER   

           CITY SEX   AGE  DRIVERS_LICENSE_CLASS TRAVEL_DIRECTION  \
0       CHICAGO   M  45.0                      D                W   
1       CHICAGO   M  45.0                      D                W   
2  WEST CHICAGO   M  69.0                     DM                W   
3  WEST CHICAGO   M  69.0                     DM                W   
4           NaN   X   NaN                    NaN                E   

         MANEUVER  
0  STRAIGHT AHEAD  
1  STRAIGHT AHEAD  
2  STRAIGHT AHEAD  
3  STRAIGHT AHEAD  
4   TURNING RIGHT  

[5 rows x 21 columns]

I can already tell we'll have way to many columns once I one-hot encode these. Let's go back and see what other columns we can drop.

In [42]:
merged_crashes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3821303 entries, 0 to 3821302
Data columns (total 21 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   CRASH_RECORD_ID          object 
 1   POSTED_SPEED_LIMIT       int64  
 2   WEATHER_CONDITION        object 
 3   LIGHTING_CONDITION       object 
 4   ROADWAY_SURFACE_COND     object 
 5   ROAD_DEFECT              object 
 6   PRIM_CONTRIBUTORY_CAUSE  object 
 7   STREET_DIRECTION         object 
 8   NUM_UNITS                int64  
 9   CRASH_HOUR               int64  
 10  CRASH_DAY_OF_WEEK        int64  
 11  CRASH_MONTH              int64  
 12  LATITUDE                 float64
 13  LONGITUDE                float64
 14  PERSON_TYPE              object 
 15  CITY                     object 
 16  SEX                      object 
 17  AGE                      float64
 18  DRIVERS_LICENSE_CLASS    object 
 19  TRAVEL_DIRECTION         object 
 20  MANEUVER                 object 
dtypes: float

This is a start, but there will still be too many columns once the categorical features are preprocessed. Let's refine these features further.

'STREET_DIRECTION' and 'TRAVEL_DIRECTION' seem redundant; let's take a closer look.

In [43]:
merged_crashes['STREET_DIRECTION'].value_counts()

STREET_DIRECTION
W    1375616
S    1284931
N     900965
E     259777
Name: count, dtype: int64

In [44]:
merged_crashes['TRAVEL_DIRECTION'].value_counts()

TRAVEL_DIRECTION
N          895177
S          876826
W          806046
E          790611
UNKNOWN    188521
SE          51437
NW          46336
SW          38910
NE          38139
Name: count, dtype: int64

It seems like travel direction is more pertinent than street direction, so we'll drop the streets. Next, let's investigate 'DRIVERS_LICENSE_CLASS' more closely.

In [45]:
merged_crashes['DRIVERS_LICENSE_CLASS'].value_counts()

DRIVERS_LICENSE_CLASS
D     1631638
A       63603
C       50777
B       49613
DM      27448
       ...   
AZ          1
TA          1
C2          1
TR          1
LT          1
Name: count, Length: 281, dtype: int64

The distribution here is pretty skewed, the second most frequent class is less than 4% of the most frequent, so I don't think it will be much help. Let's check out 'ROAD_DEFECT' as well.

In [46]:
merged_crashes['ROAD_DEFECT'].value_counts()

ROAD_DEFECT
NO DEFECTS           3107058
UNKNOWN               655439
OTHER                  18539
WORN SURFACE           15601
RUT, HOLES             15499
SHOULDER DEFECT         6612
DEBRIS ON ROADWAY       2555
Name: count, dtype: int64

Similar story here; after the less-than-useful classes of 'unknown' and 'other', the first specific one is 'worn surface', but it only amounts to 0.6% of the 'no defects' class, so this column will likely only hinder our model.

In [47]:
merged_crashes['ROADWAY_SURFACE_COND'].value_counts()

ROADWAY_SURFACE_COND
DRY                2853708
WET                 538994
UNKNOWN             274972
SNOW OR SLUSH       119147
ICE                  24736
OTHER                 8807
SAND, MUD, DIRT        939
Name: count, dtype: int64

In [48]:
final_columns_to_drop = ['CRASH_RECORD_ID', 'STREET_DIRECTION', 'DRIVERS_LICENSE_CLASS', 'ROAD_DEFECT']
merged_crashes = merged_crashes.drop(final_columns_to_drop, axis=1)
merged_crashes.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3821303 entries, 0 to 3821302
Data columns (total 17 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   POSTED_SPEED_LIMIT       int64  
 1   WEATHER_CONDITION        object 
 2   LIGHTING_CONDITION       object 
 3   ROADWAY_SURFACE_COND     object 
 4   PRIM_CONTRIBUTORY_CAUSE  object 
 5   NUM_UNITS                int64  
 6   CRASH_HOUR               int64  
 7   CRASH_DAY_OF_WEEK        int64  
 8   CRASH_MONTH              int64  
 9   LATITUDE                 float64
 10  LONGITUDE                float64
 11  PERSON_TYPE              object 
 12  CITY                     object 
 13  SEX                      object 
 14  AGE                      float64
 15  TRAVEL_DIRECTION         object 
 16  MANEUVER                 object 
dtypes: float64(3), int64(5), object(9)
memory usage: 495.6+ MB


Before we begin preprocessing this merged DataFrame, we need to go through and eliminate the rows that contain 'unknown', 'other', or 'not reported' values as well as those that represent only a tiny fraction of the feature's value counts. This will help reduce the large amount of records we have and the amount of columns that will result after one-hot encoding.

In [49]:
merged_crashes['POSTED_SPEED_LIMIT'].value_counts()

POSTED_SPEED_LIMIT
30    2915125
35     273505
25     214449
20     127854
15      99682
10      63520
40      44133
45      29661
0       28913
5       15887
55       4147
50       1232
3         977
9         400
39        374
99        291
60        252
1         125
24        122
2          84
65         82
32         75
33         62
34         60
36         40
11         40
26         35
6          26
7          22
14         18
70         15
31         14
29         11
22         10
12          9
23          8
8           8
38          8
4           7
44          6
18          6
63          4
62          2
49          1
46          1
Name: count, dtype: int64

Our first rows to drop are in the 'POSTED_SPEED_LIMIT' column. There are a lot of unusual speed limit values that aren't divisible by 5 (like 39, 24, 32). Sometimes these are done on purpose to catch the driver's attention, but at the end of the day, these will probably add more noise than anything else to our dataset.

In [50]:
import numpy as np
realistic_speed_limits = list(np.arange(0,75,5))
merged_crashes = merged_crashes[merged_crashes['POSTED_SPEED_LIMIT'].isin(realistic_speed_limits)]
merged_crashes['POSTED_SPEED_LIMIT'].value_counts()

POSTED_SPEED_LIMIT
30    2915125
35     273505
25     214449
20     127854
15      99682
10      63520
40      44133
45      29661
0       28913
5       15887
55       4147
50       1232
60        252
65         82
70         15
Name: count, dtype: int64

In [51]:
merged_crashes.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3818457 entries, 0 to 3821302
Data columns (total 17 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   POSTED_SPEED_LIMIT       int64  
 1   WEATHER_CONDITION        object 
 2   LIGHTING_CONDITION       object 
 3   ROADWAY_SURFACE_COND     object 
 4   PRIM_CONTRIBUTORY_CAUSE  object 
 5   NUM_UNITS                int64  
 6   CRASH_HOUR               int64  
 7   CRASH_DAY_OF_WEEK        int64  
 8   CRASH_MONTH              int64  
 9   LATITUDE                 float64
 10  LONGITUDE                float64
 11  PERSON_TYPE              object 
 12  CITY                     object 
 13  SEX                      object 
 14  AGE                      float64
 15  TRAVEL_DIRECTION         object 
 16  MANEUVER                 object 
dtypes: float64(3), int64(5), object(9)
memory usage: 524.4+ MB


In [52]:
merged_crashes.shape

(3818457, 17)

We have a ton of rows so we can afford to drop those which aren't complete and contain values such as 'unknown', 'other', and 'not reported'. This will help trim our data for better, more complete predicting power and also help it run a little faster.

In [53]:
unhelpful_strings = ['UNKNOWN', 'OTHER', 'NOT REPORTED']
#merged_crashes = merged_crashes[merged_crashes.apply(lambda x: not x.str.contains(unhelpful_strings).any(), axis=1)]
merged_crashes = merged_crashes[~merged_crashes.isin(unhelpful_strings).any(axis=1)]
merged_crashes.info()

<class 'pandas.core.frame.DataFrame'>
Index: 3286223 entries, 0 to 3821302
Data columns (total 17 columns):
 #   Column                   Dtype  
---  ------                   -----  
 0   POSTED_SPEED_LIMIT       int64  
 1   WEATHER_CONDITION        object 
 2   LIGHTING_CONDITION       object 
 3   ROADWAY_SURFACE_COND     object 
 4   PRIM_CONTRIBUTORY_CAUSE  object 
 5   NUM_UNITS                int64  
 6   CRASH_HOUR               int64  
 7   CRASH_DAY_OF_WEEK        int64  
 8   CRASH_MONTH              int64  
 9   LATITUDE                 float64
 10  LONGITUDE                float64
 11  PERSON_TYPE              object 
 12  CITY                     object 
 13  SEX                      object 
 14  AGE                      float64
 15  TRAVEL_DIRECTION         object 
 16  MANEUVER                 object 
dtypes: float64(3), int64(5), object(9)
memory usage: 451.3+ MB


In [54]:
#merged_crashes['ROADWAY_SURFACE_COND'].drop(['SAND, MUD, DIRT'], axis=0, inplace=True)
#merged_crashes['ROADWAY_SURFACE_COND'] = ~merged_crashes['ROADWAY_SURFACE_COND'].str.contains('SAND, MUD, DIRT')
#substring = 'SAND, MUD, DIRT'
filter = merged_crashes['ROADWAY_SURFACE_COND'].str.contains('SAND, MUD, DIRT')
merged_crashes = merged_crashes[~filter]
merged_crashes['ROADWAY_SURFACE_COND'].value_counts()

ROADWAY_SURFACE_COND
DRY              2653299
WET               502471
SNOW OR SLUSH     108198
ICE                21492
Name: count, dtype: int64

Now we'll narrow our rows to only show crashes that occured in the city of Chicago:

In [55]:
merged_crashes = merged_crashes[merged_crashes['CITY'] == 'CHICAGO']
merged_crashes['CITY'].value_counts()

CITY
CHICAGO    1764860
Name: count, dtype: int64

Now that we know that worked and helped narrow down the rows to a more pertinent subset, we can drop this column since it only contains one entry:

In [56]:
merged_crashes.drop('CITY', axis=1, inplace=True)

We also need to filter out all rows with the 'passenger' class in the 'PERSON_TYPE' column since it is far less likely that they were a contributing cause to the accident than a driver, pedestrian, or cyclist was. We'll also remove the 'non-motor vehicle' and 'non-contact vehcile' rows since they are so few and far between that including them as one-hot encoded columns will likely just hinder our model.

In [57]:
unhelpful_strings = ['PASSENGER', 'NON-MOTOR VEHICLE', 'NON-CONTACT VEHCILE']
merged_crashes = merged_crashes[~merged_crashes['PERSON_TYPE'].str.contains('|'.join(unhelpful_strings))]
merged_crashes['PERSON_TYPE'].value_counts()

PERSON_TYPE
DRIVER                 1391006
PEDESTRIAN               33218
BICYCLE                  17923
NON-CONTACT VEHICLE        234
Name: count, dtype: int64

Next, we'll remove some strange negative numbers from the 'AGE' column:

In [58]:
#merged_crashes = merged_crashes.drop(merged_crashes[merged_crashes['AGE'] < 0].index, how='any')
#merged_crashes = merged_crashes['AGE'].drop(merged_crashes['AGE'] < 0)
merged_crashes = merged_crashes[merged_crashes['AGE'] >= 0]
merged_crashes['AGE'].sort_values(ascending=True)

3038791      0.0
2714031      0.0
2714032      0.0
1510897      0.0
2545014      0.0
           ...  
1350752    109.0
200079     109.0
998681     110.0
998680     110.0
2550268    110.0
Name: AGE, Length: 1360770, dtype: float64

Let's also drop the classes from 'MANUEVER' that total fewer than 1,000 since there are 27 classes here. We'll also get rid of NaN's that are described as "UNKNOWN/NA" here.

In [59]:
#First, the classes with fewer than 1,000 entries
value_counts = merged_crashes['MANEUVER'].value_counts()
merged_crashes = merged_crashes[merged_crashes['MANEUVER'].isin(value_counts[value_counts >= 1000].index)]
#Now the specific 'unknown/na' class
merged_crashes = merged_crashes[~merged_crashes['MANEUVER'].str.contains('UNKNOWN/NA')]
merged_crashes['MANEUVER'].value_counts()

MANEUVER
STRAIGHT AHEAD                        756311
SLOW/STOP IN TRAFFIC                  126222
TURNING LEFT                          104934
PARKED                                 61085
TURNING RIGHT                          50413
BACKING                                42772
PASSING/OVERTAKING                     34536
CHANGING LANES                         28588
ENTERING TRAFFIC LANE FROM PARKING     18851
MERGING                                10796
STARTING IN TRAFFIC                    10028
U-TURN                                  9191
AVOIDING VEHICLES/OBJECTS               6267
ENTER FROM DRIVE/ALLEY                  6136
SKIDDING/CONTROL LOSS                   5386
LEAVING TRAFFIC LANE TO PARK            4990
SLOW/STOP - LEFT TURN                   3231
DRIVING WRONG WAY                       1981
PARKED IN TRAFFIC LANE                  1893
SLOW/STOP - RIGHT TURN                  1862
SLOW/STOP - LOAD/UNLOAD                 1556
NEGOTIATING A CURVE                     1430
N

Finally, let's check and see how the date & time data looks. The [data dictionary tells us](https://data.cityofchicago.org/Transportation/Traffic-Crashes-Crashes/85ca-t3if/about_data) that Sunday=1, Monday=2, etc.

In [60]:
merged_crashes['CRASH_DAY_OF_WEEK'].value_counts()

CRASH_DAY_OF_WEEK
6    214916
5    190092
3    188682
4    188315
7    183845
2    176736
1    145873
Name: count, dtype: int64

So, this columns appears as a numerical one, but it's really a categorical column that will need to be one-hot encoded. The same goes for 'CRASH_MONTH':

In [61]:
merged_crashes['CRASH_MONTH'].value_counts()

CRASH_MONTH
10    120868
9     113600
12    110461
8     109769
11    107717
7     106111
5     105815
3     105160
6     104840
4     103804
1     101555
2      98759
Name: count, dtype: int64

However, we'll keep the hour of day as a numerical column [since it can serve as one](https://stats.stackexchange.com/a/245887/378499) and it would also help keep dimensionality down.

# preparing target column

The goal of this study is to explain the underlying causes in vehicle crashes, our target column will be 'PRIM_CONTIBUTORY_CAUSE'. The only obstacle here is the amount of classes this column contains:

In [62]:
merged_crashes['PRIM_CONTRIBUTORY_CAUSE'].value_counts()

PRIM_CONTRIBUTORY_CAUSE
UNABLE TO DETERMINE                                                                 360273
FAILING TO YIELD RIGHT-OF-WAY                                                       204131
FOLLOWING TOO CLOSELY                                                               172464
FAILING TO REDUCE SPEED TO AVOID CRASH                                               75528
IMPROPER OVERTAKING/PASSING                                                          68475
IMPROPER TURNING/NO SIGNAL                                                           53608
IMPROPER LANE USAGE                                                                  51058
IMPROPER BACKING                                                                     40941
DISREGARDING TRAFFIC SIGNALS                                                         39556
NOT APPLICABLE                                                                       38624
DRIVING SKILLS/KNOWLEDGE/EXPERIENCE                               

In [63]:
len(merged_crashes['PRIM_CONTRIBUTORY_CAUSE'].value_counts())

40

While we can drop the 'UNABLE TO DETERMINE' and 'NOT APPLICABLE' classes, that would still leave 38 classes, which is not going to lead to any meaningful results.

My solution is to generalize these causes into two categories with the end goal for the stakeholder in mind, which is reducing traffic accidents. Specifically, I'll group these into those causes which show the driver was at fault and those which don't, like so:




In [64]:
#First, we'll drop the unhelpful classes
unhelpful_strings = ['UNABLE TO DETERMINE', 'NOT APPLICABLE']
merged_crashes = merged_crashes[~merged_crashes['PRIM_CONTRIBUTORY_CAUSE'].str.contains('|'.join(unhelpful_strings))]
merged_crashes['PRIM_CONTRIBUTORY_CAUSE'].value_counts()

PRIM_CONTRIBUTORY_CAUSE
FAILING TO YIELD RIGHT-OF-WAY                                                       204131
FOLLOWING TOO CLOSELY                                                               172464
FAILING TO REDUCE SPEED TO AVOID CRASH                                               75528
IMPROPER OVERTAKING/PASSING                                                          68475
IMPROPER TURNING/NO SIGNAL                                                           53608
IMPROPER LANE USAGE                                                                  51058
IMPROPER BACKING                                                                     40941
DISREGARDING TRAFFIC SIGNALS                                                         39556
DRIVING SKILLS/KNOWLEDGE/EXPERIENCE                                                  38568
WEATHER                                                                              23197
DISREGARDING STOP SIGN                                            

In [65]:
driver_at_fault = ['FAILING TO YIELD RIGHT-OF-WAY', 'FOLLOWING TOO CLOSELY', 'FAILING TO REDUCE SPEED TO AVOID CRASH',
                    'IMPROPER OVERTAKING/PASSING', 'IMPROPER TURNING/NO SIGNAL', 'IMPROPER LANE USAGE',
                    'IMPROPER BACKING', 'DISREGARDING TRAFFIC SIGNALS', 'DRIVING SKILLS/KNOWLEDGE/EXPERIENCE',
                    'DISREGARDING STOP SIGN', 'OPERATING VEHICLE IN ERRATIC, RECKLESS, CARELESS, NEGLIGENT OR AGGRESSIVE MANNER',
                    'DISTRACTION - FROM INSIDE VEHICLE', 'UNDER THE INFLUENCE OF ALCOHOL/DRUGS (USE WHEN ARREST IS EFFECTED)',
                    'DRIVING ON WRONG SIDE/WRONG WAY', 'DISREGARDING OTHER TRAFFIC SIGNS', 'EXCEEDING SAFE SPEED FOR CONDITIONS',
                    'EXCEEDING AUTHORIZED SPEED LIMIT', 'CELL PHONE USE OTHER THAN TEXTING', 'DISREGARDING ROAD MARKINGS',
                    'TURNING RIGHT ON RED', 'HAD BEEN DRINKING (USE WHEN ARREST IS NOT MADE)',
                    'DISTRACTION - OTHER ELECTRONIC DEVICE (NAVIGATION DEVICE, DVD PLAYER, ETC.)', 'TEXTING', 'RELATED TO BUS STOP',
                    'DISREGARDING YIELD SIGN', 'PASSING STOPPED SCHOOL BUS', 'BICYCLE ADVANCING LEGALLY ON RED LIGHT',
                    'MOTORCYCLE ADVANCING LEGALLY ON RED LIGHT']

In [66]:
driver_not_at_fault = ['WEATHER', 'EQUIPMENT - VEHICLE CONDITION', 'PHYSICAL CONDITION OF DRIVER',
                       'VISION OBSCURED (SIGNS, TREE LIMBS, BUILDINGS, ETC.)', 'DISTRACTION - FROM OUTSIDE VEHICLE',
                       'ROAD ENGINEERING/SURFACE/MARKING DEFECTS', 'ROAD CONSTRUCTION/MAINTENANCE',
                       'EVASIVE ACTION DUE TO ANIMAL, OBJECT, NONMOTORIST']

In [67]:
#engineer new feature with condensed causes
merged_crashes['TARGET'] = merged_crashes['PRIM_CONTRIBUTORY_CAUSE'].apply(lambda x: 1 if x in driver_at_fault else 0)
merged_crashes['TARGET'].value_counts()

TARGET
1    825529
0     64033
Name: count, dtype: int64

In [68]:
merged_crashes['TARGET'].value_counts(normalize=True)

TARGET
1    0.928017
0    0.071983
Name: proportion, dtype: float64

It looks like we will have to contend with some substantial - but not extreme - target class imbalance when fitting our models. Just to be sure, let's take a random sample of our merged_crashes DataFrame to make sure things look right:

In [69]:
merged_crashes.sample(n=50, random_state=210)

POSTED_SPEED_LIMIT WEATHER_CONDITION      LIGHTING_CONDITION  \
1179131                  20      BLOWING SNOW                DAYLIGHT   
2842147                  30             CLEAR                    DUSK   
3350549                  30             CLEAR                DAYLIGHT   
2089100                  30             CLEAR                DAYLIGHT   
1922676                  30             CLEAR  DARKNESS, LIGHTED ROAD   
1239459                  15        SLEET/HAIL                DAYLIGHT   
3084427                  30             CLEAR                DAYLIGHT   
2265765                  30              RAIN                DAYLIGHT   
3141547                  20              SNOW                DAYLIGHT   
2876459                  30             CLEAR                DAYLIGHT   
463950                   30             CLEAR                DAYLIGHT   
2228816                  35             CLEAR                DAYLIGHT   
3249472                  35             CLEAR                DAYLIGHT   
944428                   30             CLEAR                DAYLIGHT   
1937800                  30             CLEAR                DAYLIGHT   
2845612                  30             CLEAR                DAYLIGHT   
2233356                  30   CLOUDY/OVERCAST                DAYLIGHT   
3809785                  30             CLEAR  DARKNESS, LIGHTED ROAD   
1048139                  30             CLEAR  DARKNESS, LIGHTED ROAD   
2734624                  30             CLEAR                DAYLIGHT   
3327728                  30             CLEAR                DAYLIGHT   
1538473                  30             CLEAR                    DUSK   
2843509                  30              RAIN                DAYLIGHT   
3297726                  25   CLOUDY/OVERCAST                DAYLIGHT   
934154                   30   CLOUDY/OVERCAST                    DUSK   
2159509                  30             CLEAR  DARKNESS, LIGHTED ROAD   
777935                   30             CLEAR                DAYLIGHT   
3420977                  20             CLEAR  DARKNESS, LIGHTED ROAD   
3230919                  30             CLEAR                DAYLIGHT   
836894                   30             CLEAR                DARKNESS   
1748459                  30             CLEAR                DAYLIGHT   
950752                   30             CLEAR                DAYLIGHT   
2165450                  30             CLEAR                DARKNESS   
659201                   30             CLEAR                DAYLIGHT   
1873590                  30             CLEAR                DAYLIGHT   
1101163                  35             CLEAR                DAYLIGHT   
397128                   20             CLEAR                DAYLIGHT   
343251                   30             CLEAR  DARKNESS, LIGHTED ROAD   
3520182                  30             CLEAR                DAYLIGHT   
2907283                  30             CLEAR                DAYLIGHT   
3773165                  30             CLEAR                DAYLIGHT   
881543                   30              RAIN                DAYLIGHT   
2478368                  30             CLEAR                DAYLIGHT   
583582                   30             CLEAR                DAYLIGHT   
3098013                  30              RAIN  DARKNESS, LIGHTED ROAD   
435642                   30             CLEAR                DAYLIGHT   
247585                   35              RAIN                DAYLIGHT   
279701                   30             CLEAR  DARKNESS, LIGHTED ROAD   
1818999                  30             CLEAR  DARKNESS, LIGHTED ROAD   
3753782                  35              RAIN  DARKNESS, LIGHTED ROAD   

        ROADWAY_SURFACE_COND  \
1179131                  ICE   
2842147                  DRY   
3350549                  DRY   
2089100                  DRY   
1922676                  DRY   
1239459                  ICE   
3084427                  DRY   
2265765                  WET 

Now we can drop the original 'PRIM_CONTRIBUTORY_CAUSE' column to avoid the multicollinearity that would arise (plus the exploded categorical columns). We can also drop the 'CITY' column since every accident here will have the same value ('CHICAGO').

In [70]:
merged_crashes.drop('PRIM_CONTRIBUTORY_CAUSE', axis=1, inplace=True)
merged_crashes.info()

<class 'pandas.core.frame.DataFrame'>
Index: 889562 entries, 0 to 3821256
Data columns (total 16 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   POSTED_SPEED_LIMIT    889562 non-null  int64  
 1   WEATHER_CONDITION     889562 non-null  object 
 2   LIGHTING_CONDITION    889562 non-null  object 
 3   ROADWAY_SURFACE_COND  889562 non-null  object 
 4   NUM_UNITS             889562 non-null  int64  
 5   CRASH_HOUR            889562 non-null  int64  
 6   CRASH_DAY_OF_WEEK     889562 non-null  int64  
 7   CRASH_MONTH           889562 non-null  int64  
 8   LATITUDE              884826 non-null  float64
 9   LONGITUDE             884826 non-null  float64
 10  PERSON_TYPE           889562 non-null  object 
 11  SEX                   889508 non-null  object 
 12  AGE                   889562 non-null  float64
 13  TRAVEL_DIRECTION      889562 non-null  object 
 14  MANEUVER              889562 non-null  object 
 15  TARG

Wait, I think you do the stuff below in the next notebook.

In [71]:
#from sklearn.preprocessing import OneHotEncoder
#encoder = OneHotEncoder()
#categorical_columns = ['WEATHER_CONDITION', 'LIGHTING_CONDITION', 'ROADWAY_SURFACE_COND', 'CRASH_DAY_OF_WEEK',
#                       'CRASH_MONTH', 'PERSON_TYPE', 'SEX', 'TRAVEL_DIRECTION', 'MANEUVER']